packages/extension/src/background/types.ts

```diff
@@ -4,89 +4,137 @@
-  origin: string;
-}
+type IsNull<T, K extends keyof T> = { [K1 in Exclude<keyof T, K>]: T[K1] } & T[K] extends null ? K : never
+type NullKeys<T> = { [K in keyof T]: IsNull<T, K> }[keyof T]
+export type MessageTypesWithNullRequest = NullKeys<RequestTypes>
 
-export interface MessageAuthorizeApprove {
+export interface TransportRequestMessage<TMessageType extends MessageTypes> {
   id: string;
+  message: TMessageType;
+  origin: 'page' | 'popup';
+  request: RequestTypes[TMessageType];
 }
 
-export interface MessageAuthorizeReject {
-  id: string;
+export interface RequestAuthorizeTab {
+  origin: string;
 }
 
-export interface MessageRequest {
+export interface RequestAuthorizeApprove {
   id: string;
-  message: MessageTypes;
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  request: any;
 }
 
-export interface MessageResponse {
-  error?: string;
+export interface RequestAuthorizeReject {
   id: string;
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  response?: any;
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  subscription?: any;
 }
 
-export interface MessageAccountCreate {
+export type RequestAuthorizeRequests = null;
+
+export type RequestAuthorizeSubscribe = null;
+
+export interface RequestAccountCreate {
   name: string;
   password: string;
   suri: string;
   type?: KeypairType;
 }
 
-export interface MessageAccountEdit {
+export interface RequestAccountEdit {
   address: string;
   name: string;
 }
 
-export interface MessageAccountForget {
+export interface RequestAccountForget {
   address: string;
 }
 
-export interface MessageExtrinsicSignApprove {
+export type RequestAccountList = null;
+
+export type RequestAccountSubscribe = null;
+
+export type RequestExtrinsicSign = SignerPayload;
+
+export interface RequestSigningApprove {
   id: string;
   password: string;
 }
 
-export interface MessageExtrinsicSignCancel {
+export interface RequestSigningCancel {
   id: string;
 }
 
-export type MessageExtrinsicSign = SignerPayload;
-
-export interface MessageExtrinsicSignResponse {
-  id: string;
-  signature: string;
-}
+export type RequestSigningRequests = null;
 
-export interface MessageSeedCreateResponse {
-  address: string;
-  seed: string;
-}
+export type RequestSigningSubscribe = null;
 
-export interface MessageSeedCreate {
+export interface RequestSeedCreate {
   length?: 12 | 24;
   type?: KeypairType;
 }
 
-export interface MessageSeedValidate {
+export interface RequestSeedValidate {
   suri: string;
   type?: KeypairType;
 }
 
-export interface MessageSeedValidateResponse {
+// Responses
+
+export type ResponseTypes = {
+  [MessageType in keyof RequestSignatures]: RequestSignatures[MessageType][1]
+};
+
+export interface TransportResponseMessage<TMessageType extends MessageTypes> {
+  error?: string;
+  id: string;
+  response?: ResponseTypes[TMessageType];
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  subscription?: any;
```

Comment:
Which type is `subscription` meant to be?

Reply:
Good question, it actually doesn't have to be `any` here.

In case of `accounts.subscribe`, `authorize.subscribe`, `signing.subscribe` : `subscription` (subscription message actually) will be respectively of type `KeyringJson[]`, `AuthorizeRequest[]`, `SigningRequest[]`.

Removed this `any` and the one in the `sendMessage` subscription callback (so we have typecheck for the callback now as well)

Reply:
Brilliant, thanks.